### PR TITLE
Update matrix inverse root via CANS

### PR DESCRIPTION
**Pull request by skyw**
Disallow bf16 compute for CANS

Saved couple of mm by moving things out of main loop. Slight readability loss.

@xsgxlz provided test case showing that bf16 compute is insufficient for CANS. Choices are to silently uses higher precision or raise error. decided to raise error.

```
import torch

from emerging_optimizers.soap.matrix_root_inverse_utils import (
    mat_root_inv_via_scaled_cans,
)

device = "cuda"
size = 1024

generator = torch.Generator(device=device).manual_seed(0)
x = torch.randn(size, size, generator=generator, device=device)

matrix = (x @ x.T) / size
matrix.diagonal().add_(1e-6)

with torch.no_grad():
    for precision in ("medium", "high"):
        inverse_root = mat_root_inv_via_scaled_cans(
            matrix,
            fp32_matmul_prec=precision,
        )

        print(f"{precision}:")
        print("  finite:", torch.isfinite(inverse_root).sum().item())
        print("  NaN:", torch.isnan(inverse_root).sum().item())
        print("  Inf:", torch.isinf(inverse_root).sum().item())
```

### CODE REVIEW

**Comment by xsgxlz** [comment]
Thanks for the update!

Also, please consider this version, which saves one more mm analytically

```python
    inf_norm = torch.linalg.matrix_norm(x, ord=float("inf"), dim=(-2, -1), keepdim=True).clamp_min_(eps)
    y = x / inf_norm

    # --- NEW INITIALIZATION (Step 0) ---
    beta, alpha = _CANS_COEFFS[0]
    
    # Z_1 = beta * I + alpha * Y_0 (0 MMs)
    z = y * alpha
    z.diagonal(dim1=-2, dim2=-1).add_(beta)
    
    # Y_1 = beta * Y_0 + alpha * Y_0^2 (1 MM)
    y = torch.baddbmm(y, y, y, beta=beta, alpha=alpha)

    # --- COUPLED UPDATES (Steps 1 to 8) ---
    for beta, alpha in _CANS_COEFFS[1:-1]:
        p = z @ y
        # We can tuple-unpack to ensure both use the 'old' z and y
        z, y = (
            torch.baddbmm(z, p, z, beta=beta, alpha=alpha),
            torch.baddbmm(y, y, p, beta=beta, alpha=alpha),
        )

    # --- FINAL UPDATE (Step 9) ---
    beta, alpha = _CANS_COEFFS[-1]
    p = z @ y
    z = torch.baddbmm(z, p, z, beta=beta, alpha=alpha)

    # Apply scaling and symmetrize
    z.mul_(torch.rsqrt(inf_norm))
    result = (z + z.mT) * 0.5
    return result if is_batched else result.squeeze(0)
```

File: emerging_optimizers/soap/matrix_root_inverse_utils.py [hunk 1]
```diff
@@ -68,25 +70,28 @@ def mat_root_inv_via_scaled_cans(
         raise TypeError(f"x must be a square matrix or batch of square matrices, got shape {tuple(x.shape)}")
     if x.dtype != torch.float32:
         raise TypeError(f"x must be in float32, got {x.dtype}")
+    if torch.get_float32_matmul_precision() == "medium":
```
Comment (mkhona-nvidia) [inline emerging_optimizers/soap/matrix_root_inverse_utils.py:73]: this function would be wrapped with a fp32 mamtul decorator, would check via torch only check the local value of this variable?

**Review by mkhona-nvidia — APPROVED** [review]

File: emerging_optimizers/soap/matrix_root_inverse_utils.py (hunk 1 above)
Comment (skyw) [reply emerging_optimizers/soap/matrix_root_inverse_utils.py:73]: The value is global in torch.
